gcode-svg.1
.TH GCODE-TEXT 1
.SH NAME
gcode-svg \- render SVG as gcode
.SH SYNOPSIS
gcode-svg [options] [input files] ...
.SH DESCRIPTION
.PP
\fIGcode-svg\fP renders stroked SVG elements to gcode. Each SVG path is
drawn using selectable speed and feed values based on a
color-to-parameter mapping file.
.SH USAGE
.PP
An un-flagged argument is treated as an input file.
.PP
Options are as follows:
.TP
.B "--help,--usage"
Print a help/usage message and exit.
.TP
.BI "-V,--version"
Prints the version number and immediately exits.
.TP
.BI "-i,--inch"
Use inch units in the generated file. This is the default.
.TP
.BI "-m,--mm"
Use millimeter units in the generated file.
.TP
.BI "-x,--x-off"
Offset drawing in X by this much from the device origin. This value is
in the output units.
.TP
.BI "-y,--y-off"
Offset drawing in y by this much from the device origin. This value is
in the output units.
.TP
.BI "--tesselate"
Tesselate splines even when the output device supports them natively.
.TP
.BI "--sheer " sheer
Set the sheer amount for oblique glyphs.
.TP
.BI "-f,--flatness " flatness
Set the spline decomposition tolerance in output units. The default is
0.001.
.TP
.BI "--feed " feed
Set the feed rate. Note that this value depends on the units in use.
.TP
.BI "--speed " speed
Set the spindle speed. Nominally in RPM, check your device documentation
for non-rotating tools to understand what it does.
.TP
.BI "-d,--device " device-spec
Specifies a device file. This is a JSON file containing
definitions for drawing to the target. See the DEVICE SPEC section
below.
.TP
.BI "-S,--settings " setting,...
Specifies a comma-separated list of device-specific settings. See
the "settings" and "setting-values" items in the DEVICE SPEC section below.
.TP
.BI "-o,--output " output
Specifies the output file name.
.TP
.BI "-C,--config-dir " directory
Specifies an additional path to device configuration files. This
argument may be repeated.
.TP
.BI "-p,--params " parameter-file
Specifies the parameters used during conversion. See the PARAMETERS
section below
.SH DEVICE SPEC
.PP
This file contains device-specific output customization, providing
control over the strings used to drive the target. This JSON file
should contain a single dictionary with keys mapping to string
values. Each value can contain embedded newlines as necessary;
gcode-svg doesn't emit any newlines itself. If not provided in the
configuration file (or if no configuration file is provided), the
default values shown below will be used instead.
.TP
.BI "start " "G90\\\\\\\\nG17\\\\\\\\n"
Specifies a string emitted before any drawing is done to set up the
target.
.TP
.BI "settings " "\"\""
Specifies a C-style printf format for outputting device-specific
settings. If the string is empty, this element is skipped.
.TP
.BI "setting-values " "\"\""
Specifies an array of strings, one for each %s value present in the
"settings" value. These can be replaced on the command line using
the --settings option.
.TP
.BI "inch " "G20\\\\\\\\n"
Specifies the string to place the device in "inch" mode.
.TP
.BI "mm " "G21\\\\\\\\n"
Specifies the string to place the device in "millimeter" mode.
.TP
.BI "move " "G00 X%f Y%f\\\\\\\\n"
Specifies a C-style printf format for repositioning the output without
drawing. This format should contain two %f instances, the first is
used for the X coordinate, the second for Y.
.TP
.BI "draw " "G01 X%f Y%f F%f S%f\\\\\\\\n"
Specifies a C-style printf format for drawing a straight line. This
format should contain two to four %f instances, the first is used for the X
coordinate, the second for Y, followed by an (optional) feed rate and
(optional) spindle speed.
.TP
.HI "z-move " "G01 Z%f F%f S%f\\\\\\\\n"
Specifies a C-style printf format for moving the tool in the Z
axis. This
format should contain one to three %f instances, followed by an
(optional) feed rate and (optional) spindle speed.
.TP
.BI "feed " "true or false"
Indicates whether the draw element includes a %f required for
the feed rate.
.TP
.BI "speed " "true or false"
Indicates whether the draw element includes a %f required for
the spindle speed.
.TP
.BI "y-invert " "true or false"
Indicates whether y values indicate upwards (true) or downwards
(false) motion.
.TP
.BI "stop " "M30\\\\\\\\n"
Specifies the string emitted after all drawing is done.
.PP
Gcode-text first emits the "start" string, then one of "inch" or
"mm". Then, drawing is done by emitting a sequence of "move" and
"draw" operations. Finally, drawing is finished with the "stop" string
and the output file is closed.
.PP
Here's what the default configuration looks like in JSON form:
.nf
{
    "start" : "G90\nG17\\n",
    "inch" : "G20\\n",
    "mm" : "G10\\n",
    "move" : "G00 X%f Y%f\\n",
    "draw" : "G01 X%f Y%f F%f S%f\\n",
    "feed" : true,
    "speed" : true,
    "stop" : "M30\\n"
}
.fi
.SH PARAMETER FILE
.PP
This (JSON formatted) file contains the mapping from stroke color to
feed and speed values. This allows you to construct a single SVG file
containing multiple operations that use different feed and speed
values. You can also place other configuration values in this file.
.PP
The parameter file contains a single JSON dictionary. Any long-form command line
option may be specified in this file by eliding the leading dashes and
replacing any other dashes with underscores. There are two additional
dictionary entries, the "params" entry and the "default" entry. The
"params" entry is an array of dictionaries, each dictionary contains
five keys:
.TP
.BI "order"
This number specifies the order in which elements will be drawn. This
allows the gcode file to draw things in an order different from how
they appear in the SVG file. You don't have to order things
sequentially; smaller order elements will be drawn before larger order
elements.
.TP
.BI "color"
This string specifies the color that this parameter will match in the
SVG file. Color matching is done by using the string value directly,
so an RGB value will be encoded as "#rrggbb". There's no slop in this
value; it must match exactly.
.TP
.BI "feed"
This number specifies the feed rate to use for matching elements. The
units depend upon the units of the generated gcode file.
.TP
.BI "step"
This number specifies the distance between adjacent scanlines used when
converting filled portions of the drawing.
.TP
.BI "speed"
This number specifies the spindle RPM to use for matching
elements. For tools without a spindle, check the documentation to
understand how this value is used.
.TP
.BI "name"
This string provides a name documenting the intended purpose of the
parameters. It isn't used by the software at all.
.SH AUTHOR
\fIGcode-svg\fP is the work of Keith Packard <keithp@keithp.com>.
.\"
.PP
Gcode-svg is
.br
Copyright 2023 Keith Packard.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,


debian/changelog
gcode-text (0.8-1) unstable; urgency=medium

  * Add --x-off and --y-off to gcode-svg
  * Match colors by value instead of by string in gcode-svg

 -- Keith Packard <keithp@keithp.com>  Sun, 17 May 2026 21:54:27 -0700

gcode-text (0.7-1) unstable; urgency=medium

  * Add Z motion control
  * Add scan conversion for filled areas

 -- Keith Packard <keithp@keithp.com>  Sun, 17 May 2026 14:09:45 -0700

gcode-text (0.6-2) unstable; urgency=medium

  * Make python dependencies explicit

 -- Keith Packard <keithp@keithp.com>  Sun, 31 Dec 2023 17:19:32 -0800

gcode-text (0.6-1) unstable; urgency=medium

  * Add gcode-svg
  * Add script font
  * Fix text centering in y-invert mode.
  * Add parametric template support
  * Lots of font updates
  * Switch fonts to svg format

 -- Keith Packard <keithp@keithp.com>  Sun, 31 Dec 2023 14:22:55 -0800

gcode-text (0.5-1) unstable; urgency=medium

  * Switch to python
  * Change template to JSON format
  * Add --sheer and --tesselate options

 -- Keith Packard <keithp@keithp.com>  Sun, 03 Dec 2023 17:26:02 -0800

gcode-text (0.4-1) unstable; urgency=medium

  * Add spline output support
  * Add "settings" to device configurations
  * Add --version and --oblique options

 -- Keith Packard <keithp@keithp.com>  Fri, 01 Dec 2023 13:36:47 -0800

gcode-text (0.3-1) unstable; urgency=medium

  * Add SVG output support

 -- Keith Packard <keithp@keithp.com>  Thu, 30 Nov 2023 16:47:13 -0800

gcode-text (0.2-1) unstable; urgency=medium

  * Add device config file support.

 -- Keith Packard <keithp@keithp.com>  Wed, 29 Nov 2023 17:36:28 -0800

gcode-text (0.1-1) unstable; urgency=medium

  * Initial release. (Closes: #123456)

 -- Keith Packard <keithp@keithp.com>  Wed, 29 Nov 2023 12:04:50 -0800


gcode-svg.py
#!/usr/bin/python3
#
# Copyright © 2023 Keith Packard <keithp@keithp.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
#

from __future__ import annotations
import json
import math
import sys
import argparse
import os
import numbers
import re
from typing import Any
import svgelements  # type: ignore
from svgelements import *  # type: ignore

if not os.getenv("GCODE_SKIP_PATH"):
    sys.path = ["@SHARE_DIR@"] + sys.path

from gcode_draw import *
from gcode_font import *


class SvgValues(Values):

    def __init__(self):
        super().__init__()
        self.ppi = 96.0
        self.bounds = Rect(Point(0, 0), Point(0, 0))

    def set_bounds(self, bounds: Rect) -> None:
        self.bounds = bounds


class SvgColor:
    red: int
    green: int
    blue: int
    alpha: float

    names = {
        "black": "#000000",
        "silver": "#C0C0C0",
        "gray": "#808080",
        "white": "#FFFFFF",
        "maroon": "#800000",
        "red": "#FF0000",
        "purple": "#800080",
        "fuchsia": "#FF00FF",
        "green": "#008000",
        "lime": "#00FF00",
        "olive": "#808000",
        "yellow": "#FFFF00",
        "navy": "#000080",
        "blue": "#0000FF",
        "teal": "#008080",
        "aqua": "#00FFFF",
    }

    def float_value(self, text):
        return float(text)

    def decimal_value(self, text):
        if text.endswith("%"):
            return int(round(int(text[:-1]) * 255 / 100))
        else:
            return int(text)

    def hex_value(self, text):
        if len(text) == 1:
            return int(text, 16) * 17
        return int(text, 16)

    def __init__(self, text: str):
        if text in self.names:
            text = self.names[text]
        m = re.fullmatch(r"#([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])", text)
        if m:
            self.red = self.hex_value(m.group(1))
            self.green = self.hex_value(m.group(1))
            self.blue = self.hex_value(m.group(1))
            self.alpha = 1
            return

        m = re.fullmatch(r"#([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})", text)
        if m:
            self.red = self.hex_value(m.group(1))
            self.green = self.hex_value(m.group(2))
            self.blue = self.hex_value(m.group(3))
            self.alpha = 1
            return

        m = re.fullmatch(
            r"rgb *\( *([0-9][0-9]*%?) *, *([0-9][0-9]*%?) *, *([0-9][0-9]*%?) *\)",
            text,
        )
        if m:
            self.red = self.decimal_value(m.group(1))
            self.green = self.decimal_value(m.group(2))
            self.blue = self.decimal_value(m.group(3))
            self.alpha = 1
            return

        m = re.fullmatch(
            r"rgba *\( *([0-9][0-9]*%?) *, *([0-9][0-9]*%?) *, *([0-9][0-9]*%?) *, *([0-9][0-9]*%?) *\)",
            text,
        )
        if m:
            self.red = self.decimal_value(m.group(1))
            self.green = self.decimal_value(m.group(2))
            self.blue = self.decimal_value(m.group(3))
            self.alpha = self.float_value(m.group(4))
            return

    def __eq__(self, other):
        return (
            self.red == other.red
            and self.green == other.green
            and self.blue == other.blue
            and self.alpha == other.alpha
        )

    def __str__(self):
        if self.alpha < 1:
            return "rgba(%d, %d, %d, %f)" % (
                self.red,
                self.green,
                self.blue,
                self.alpha,
            )
        else:
            return "rgb(%d, %d, %d)" % (self.red, self.green, self.blue)

    def __hash__(self):
        return hash(str(self))


class Param:
    order: int
    color: SvgColor
    feed: float
    z_feed: float | None
    speed: float
    passes: int
    name: str
    step: float

    def __init__(
        self,
        order: int,
        color: str | SvgColor,
        feed: float,
        z_feed: float | None,
        speed: float,
        passes: int,
        name: str,
        step: float,
    ):
        self.order = order
        if isinstance(color, str):
            color = SvgColor(color)
        self.color = color
        self.feed = feed
        self.z_feed = z_feed
        self.speed = speed
        self.passes = passes
        self.name = name
        self.step = step


class Params:

    params: dict[SvgColor, Param]
    default: Param
    global_default: Param = Param(
        order=1,
        color="default",
        feed=1,
        z_feed=None,
        speed=1,
        passes=1,
        name="default",
        step=0.1,
    )

    def __init__(self, json_file: str, values: SvgValues):
        self.params = {}
        self.default = self.global_default
        if json_file is not None:
            with values.config_open(json_file) as file:
                self.set_params(json.load(file), values)

    def make_param(self, param, color=None):
        # order, color and name are required parameters
        order = param["order"]
        if color is None:
            color = SvgColor(param["color"])

        name = param["name"]

        feed = self.default.feed
        z_feed = self.default.z_feed
        speed = self.default.speed
        passes = self.default.passes
        step = self.default.step

        if "feed" in param:
            feed = param["feed"]
        if "z_feed" in param:
            z_feed = param["z_feed"]
        if "speed" in param:
            speed = param["speed"]
        if "passes" in param:
            passes = param["passes"]
        if "step" in param:
            step = param["step"]

        return Param(
            order=order,
            color=color,
            feed=feed,
            z_feed=z_feed,
            speed=speed,
            passes=passes,
            name=name,
            step=step,
        )

    def set_params(self, json, values: SvgValues):
        values.handle_dict(json)
        for key, value in json.items():
            if key == "params":
                for param in value:
                    p = self.make_param(param)
                    self.params[p.color] = p
            elif key == "default":
                self.default = self.make_param(value, color="default")

    def get(self, color: str | SvgColor):
        if isinstance(color, str):
            color = SvgColor(color)
        if not color in self.params:
            print("unknown color %s" % color)
            return self.default
        return self.params[color]


def stroke_to_gcode(gcode: GCode, path: svgelements.Path, param: Param, matrix: Matrix):

    if gcode.values.verbose:
        print(
            'path using "%s" feed %f z_feed %r speed %f passes %d'
            % (param.name, param.feed, param.z_feed, param.speed, param.passes)
        )
    gcode.set_feed(param.feed)
    gcode.set_z_feed(param.z_feed)
    gcode.set_speed(param.speed)

    draw = MatrixDraw(gcode.get_draw(), matrix)

    for i in range(param.passes):
        for seg in path:
            if isinstance(seg, svgelements.Move):
                if seg.end is not None:
                    draw.move(seg.end.x, seg.end.y)
            elif isinstance(seg, svgelements.Close):
                start: svgelements.Point = seg.end
                draw.draw(start.x, start.y)
            elif isinstance(seg, svgelements.Line):
                draw.draw(seg.end.x, seg.end.y)
            elif isinstance(seg, svgelements.QuadraticBezier):
                draw.curve2(seg.control.x, seg.control.y, seg.end.x, seg.end.y)
            elif isinstance(seg, svgelements.CubicBezier):
                draw.curve(
                    seg.control1.x,
                    seg.control1.y,
                    seg.control2.x,
                    seg.control2.y,
                    seg.end.x,
                    seg.end.y,
                )
            elif isinstance(seg, svgelements.Arc):
                print("arc")


class Intercept:

    x: float
    winding: bool

    def __init__(self, x, winding):
        self.x = x
        self.winding = winding

    def __str__(self):
        return "(%f %r)" % (self.x, self.winding)

    def __lt__(self, other):
        if self.x < other.x:
            return True
        if self.x > other.x:
            return False
        if self.winding and not other.winding:
            return True
        return False


class Edge:

    x1: float
    y1: float
    x2: float
    y2: float
    winding: bool

    def __init__(self, x1, y1, x2, y2):
        assert y1 != y2
        self.winding = y1 > y2
        if y1 > y2:
            t = x1
            x1 = x2
            x2 = t
            t = y1
            y1 = y2
            y2 = t
        self.x1 = x1
        self.y1 = y1
        self.x2 = x2
        self.y2 = y2

    def x(self, y):
        if y < self.y1 or self.y2 <= y:
            return None

        y_off = self.y1
        slope = (self.x2 - self.x1) / (self.y2 - self.y1)
        x_off = self.x1
        x = (y - y_off) * slope + x_off

        return Intercept(x, self.winding)

    def __str__(self):
        return "(%f %f) (%f %f)" % (self.x1, self.y1, self.x2, self.y2)

    def __lt__(self, other):
        if self.y1 < other.y1:
            return True
        if self.y1 > other.y1:
            return False

        if self.x1 < other.x1:
            return True
        if self.x1 > other.x1:
            return False

        if self.y2 < other.y2:
            return True
        if self.y2 > other.y2:
            return False

        if self.x2 < other.x2:
            return True
        return False

    def __eq__(self, other):
        if self.x1 != other.x1:
            return False
        if self.y1 != other.y1:
            return False
        if self.x2 != other.x2:
            return False
        if self.y2 != other.y2:
            return False
        return True


class EdgeDraw(Draw):

    edges: list[Edge]

    def __init__(self):
        super().__init__()
        self.edges = []

    def start(self):
        print("EdgeDraw start")
        pass

    def draw(self, x: float, y: float):
        if y != self.last_y:
            edge = Edge(x, y, self.last_x, self.last_y)
            self.edges += [edge]
        super().draw(x, y)

    def spans(self, y):
        xs = []
        for edge in self.edges:
            x = edge.x(y)
            if x is not None:
                xs += [x]
        xs.sort()
        return xs


def scan_to_gcode(
    gcode: GCode, paths: list[svgelements.Path], param: Param, matrix: Matrix
):

    edge_draw = EdgeDraw()

    line_draw = LineDraw(edge_draw, gcode.values.flatness)

    edraw = MatrixDraw(line_draw, matrix)

    for path in paths:
        for seg in path:
            if isinstance(seg, svgelements.Move):
                if seg.end is not None:
                    edraw.move(seg.end.x, seg.end.y)
            elif isinstance(seg, svgelements.Close):
                start: svgelements.Point = seg.end
                edraw.draw(start.x, start.y)
            elif isinstance(seg, svgelements.Line):
                edraw.draw(seg.end.x, seg.end.y)
            elif isinstance(seg, svgelements.QuadraticBezier):
                edraw.curve2(seg.control.x, seg.control.y, seg.end.x, seg.end.y)
            elif isinstance(seg, svgelements.CubicBezier):
                edraw.curve(
                    seg.control1.x,
                    seg.control1.y,
                    seg.control2.x,
                    seg.control2.y,
                    seg.end.x,
                    seg.end.y,
                )
            elif isinstance(seg, svgelements.Arc):
                print("arc")
            else:
                print("skipping %r" % seg)

    if gcode.values.verbose:
        print(
            'path using "%s" feed %f z_feed %r speed %f passes %d step %f'
            % (
                param.name,
                param.feed,
                param.z_feed,
                param.speed,
                param.passes,
                param.step,
            )
        )
    gcode.set_feed(param.feed)
    gcode.set_z_feed(param.z_feed)
    gcode.set_speed(param.speed)

    draw = gcode.get_draw()

    values = gcode.values

    bounds = values.bounds

    for i in range(param.passes):
        y = bounds.top_left.y
        while y <= bounds.bottom_right.y:
            spans = edge_draw.spans(y)
            if spans:
                spans.sort()
                winding = 0
                x = bounds.top_left.x
                for span in spans:
                    if winding != 0 and x != span.x:
                        draw.move(x, y)
                        draw.draw(span.x, y)
                    x = span.x
                    if span.winding:
                        winding += 1
                    else:
                        winding -= 1
            y = y + param.step


def Args():
    parser = argparse.ArgumentParser(add_help=False, description="Convert SVG to Gcode")
    Device.args(parser)
    parser.add_argument(
        "-p", "--params", action="store", help="Parameter file name", default=None
    )
    parser.add_argument(
        "-x",
        "--x-off",
        action="store",
        help="X offset drawing from origin",
        default="0",
    )
    parser.add_argument(
        "-y",
        "--y-off",
        action="store",
        help="Y offset drawing from origin",
        default="0",
    )
    parser.add_argument("file", nargs="*", help="SVG input files")

    args = parser.parse_args()

    if args.help:
        parser.print_help()
        sys.exit(0)

    if args.version:
        print("%s" % "@VERSION@")
        sys.exit(0)

    return args


def key_svg_entry(e: tuple[Param, svgelements.Path]) -> int:
    return e[1].order


def main():

    values = SvgValues()

    args = Args()

    if args.config_dir:
        values.config_dir = args.config_dir + values.config_dir

    params = Params(args.params, values)

    values.handle_args(args)

    device = Device(values)

    output = sys.stdout
    if args.output != "-":
        output = open(args.output, "w")

    svgs = ()
    for filename in args.file:
        with open(filename) as file:
            svgs += (SVG.parse(file),)

    bounds = None
    for svg in svgs:
        x = svg.implicit_x
        y = svg.implicit_y
        w = svg.implicit_width
        h = svg.implicit_height
        this_bounds = Rect(Point(x, y), Point(x + w, y + h))
        if bounds is None:
            bounds = this_bounds
        else:
            bounds = bounds.union(this_bounds)
    values.set_bounds(bounds)

    matrix = Matrix()

    if values.mm:
        units_per_inch = 25.4
    else:
        units_per_inch = 1

    scale = units_per_inch / values.ppi

    matrix = matrix.scale(scale, scale)

    if args.x_off or args.y_off:
        matrix = matrix.translate(float(args.x_off) / scale, float(args.y_off) / scale)

    if device.y_invert and values.bounds:
        matrix = matrix.translate(
            0, values.bounds.bottom_right.y + values.bounds.top_left.y
        )
        matrix = matrix.scale(1, -1)

    gcode = GCode(output, device, values, None)

    gcode.start()

    strokes = []
    fills = {}

    for svg in svgs:
        for e in svg.elements():
            stroke = e.values["stroke"]
            fill = e.values["fill"]
            if (
                isinstance(e, svgelements.Rect)
                or isinstance(e, svgelements.Circle)
                or isinstance(e, svgelements.Ellipse)
            ):
                e = svgelements.Path(e.segments())

            if isinstance(e, svgelements.Path):
                e.approximate_arcs_with_cubics()
                if stroke != "none":
                    param = params.get(stroke)
                    strokes += [(e, param)]
                if fill != "none":
                    param = params.get(fill)
                    if param in fills:
                        fills[param] += [e]
                    else:
                        fills[param] = [e]

    strokes.sort(key=key_svg_entry)

    for stroke in strokes:
        stroke_to_gcode(gcode, stroke[0], stroke[1], matrix)

    if fills:
        for param in fills:
            paths = fills[param]
            scan_to_gcode(gcode, paths, param, matrix)

    gcode.stop()


main()
